src/passes/Print.cpp

```diff
@@ -141,37 +88,112 @@ struct TypeName {
   TypeName(Type type) : type(type) {}
 };
 
-struct HeapTypeName {
-  HeapType type;
-  HeapTypeName(HeapType type) : type(type) {}
+struct ResultTypeName {
+  Type type;
+  ResultTypeName(Type type) : type(type) {}
 };
 
+static void
+printHeapTypeName(std::ostream& os, HeapType type, bool first = true);
+
+static void printTypeName(std::ostream& os, Type type) {
+  if (type.isBasic()) {
+    os << type;
+    return;
+  }
+  if (type.isTuple()) {
+    auto sep = "";
+    for (auto t : type) {
+      os << sep;
+      sep = "_";
+      printTypeName(os, t);
+    }
+  } else if (type.isRef()) {
+    os << "ref";
+    if (type.isNullable()) {
+      os << "?";
+    }
+    os << "{";
+    printHeapTypeName(os, type.getHeapType(), false);
+    os << "}";
+  } else {
+    WASM_UNREACHABLE("unsupported print type");
+  }
+}
+
+static void printHeapTypeName(std::ostream& os, HeapType type, bool first) {
+  if (type.isBasic()) {
+    os << type;
+    return;
+  }
+  if (first) {
+    os << '$';
+  }
+  if (type.isSignature()) {
+    auto sig = type.getSignature();
+    printTypeName(os, sig.params);
+    if (first) {
+      os << "_=>_";
```

Comment:
In some place we use `_->_` and elsewhere we seem to use `_=>_`... Can we unify them?

Reply:
The idea (before this PR) is to emit `=>` for the toplevel type, and `->` for inner ones. At least for me it makes things easier to read. Sometimes those types are big...

But I don't feel strongly. I can make this always emit one or the other if you prefer - let me know which one. That would make this PR change a lot of other existing tests, however, so it might be nicer as a separate followup.

Reply:
I think unifying them wouldn't be too confusing, but I don't feel strongly either. Yeah whether we do it or not it should be a followup I think as you suggested.